File: 0494. Target Sum/Solution.py
from typing import List


class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        s = sum(nums) - abs(target)
        if s < 0 or s % 2:
            return 0
        m = s // 2

        f = [1] + [0] * m
        for x in nums:
            for i in range(m, x - 1, -1):
                f[i] += f[i - x]
        return f[m]

File: 0494. Target Sum/Solution.cpp
#include <numeric>
#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
public:
    int findTargetSumWays(vector<int>& nums, int target)
    {
        int s = reduce(nums.begin(), nums.end(), 0) - abs(target);
        if (s < 0 || s % 2 == 1)
            return 0;
        int m = s / 2;

        vector<int> f(m + 1, 0);
        f[0] = 1;
        for (int x : nums)
            for (int i = m; i >= x; i--)
                f[i] += f[i - x];
        return f[m];
    }
};
